### PR TITLE
abstract_replication_strategy: prepare options: check replication_factor ambiguity

### PR DESCRIPTION
**Pull request by bhalevy**
Using the replication_factor option with
NetworkTopologyStrategy can currently lead to surprising, counter-intuitive results.

When creating a keyspace, the option initially sets the replication factor on all DCs.

But when used in ALTER KEYSPACE queries, the legacy `replication_factor` option is applied only to
DCs added after the keyspace was created, while
existing DCs are unaffected, and if no DCs were
added, the option is silently ignored.

scylladb/scylladb#8881 suggests adding a warning
in this case, but that might not be enough
as practically, looking at our testing code base,
the common assumption is that the `replication_factor` option would apply in trivial cases, like when there's only a single data center, or in successive use
after CREATE KEYSPACE, when all data centers continue to use the same replication factor.

This change detects when the `replication_factor`
can be applied safely, or return an error if
the new `replication_factor` is ambiguous.

Allow applying the replication factor only in the following cases:
- Data centers are divided into 2 groups: configured and unconfigured DCs
- If no DC is configured, apply the replication_factor to all DCs.
- Otherwise, if all configured DCs have the same RF:
  - If there are no unconfigured DCs, just apply the new replication_factor too all DCs.
  - If there are both configured DCs (all configured the same), and one or more unconfigured DCs, apply the replication_factor to the unconfigured DCs (like it is done today) iff the replication_factor is the same as the one already configured to the presently configured DCs
- Otherwise, the replication_factor is deemed as ambiguous and an error will be returned.

Fixes scylladb/scylladb#8881

### CODE REVIEW

**Comment by bhalevy** [comment]
Cc @avikivity @tzach @vladzcloudius

**Comment by bhalevy** [comment]
https://github.com/scylladb/scylla-dtest/pull/3578 adjusts the `test_rf_expand` dtest to verify this change.

**Comment by vladzcloudius** [comment]
> Cc @avikivity @tzach @vladzcloudius

I don't understand why do you call the current behavior "confusing". It is documented: https://opensource.docs.scylladb.com/stable/cql/ddl.html#networktopologystrategy

IMO no need to change anything.

**Comment by bhalevy** [comment]
> > Cc @avikivity @tzach @vladzcloudius
> 
> I don't understand why do you call the current behavior "confusing". It is documented: https://opensource.docs.scylladb.com/stable/cql/ddl.html#networktopologystrategy
> 
> IMO no need to change anything.

The fact that it is documented has nothing to do wit it being confusing and counter-intuitive
to the point that people writing dtests were not aware of it (including myself).
Nadav was surprised by it when filing #8881.
Plus, I don't think it makes much sense from the UX perspective, even though it is indeed documented.

**Comment by bhalevy** [comment]
The most confusing part of the current behavior is the silent ignoring of the replication_factor option when there are no unconfigured datacenters.  @nyh proposed to warn about it in #8881, but users, in particularly, cloud users have no reason or ability to look for this warning, especially if the CQL query returned successfully.

**Comment by vladzcloudius** [comment]
> > > Cc @avikivity @tzach @vladzcloudius
> > 
> > 
> > I don't understand why do you call the current behavior "confusing". It is documented: https://opensource.docs.scylladb.com/stable/cql/ddl.html#networktopologystrategy
> > IMO no need to change anything.
> 
> The fact that it is documented has nothing to do with it being confusing and counter-intuitive to the point that people writing dtests were not aware of it (including myself). Nadav was surprised by it when filing #8881. Plus, I don't think it makes much sense from the UX perspective, even though it is indeed documented.

Clearly they didn't read the documentation I have referenced which is super clear.

@nyh @roydahan were you aware of that doc page? If you read it would it still be "confusing"? 

IMO there is nothing confusing about it. Nor I heard customers complaining about it either.

Changing documented user facing APIs must be well justified. I don't think this is such a case.

**Comment by vladzcloudius** [comment]
> The most confusing part of the current behavior is the silent ignoring of the replication_factor option when there are no unconfigured datacenters. @nyh proposed to warn about it in #8881, but users, in particularly, cloud users have no reason or ability to look for this warning, especially if the CQL query returned successfully.

Again, this is not confusing at all - this is a documented behavior.

**Comment by bhalevy** [comment]
> Again, this is not confusing at all - this is a documented behavior.

This is not a legal argument.
Let me rephrase that as counter-intuitive since without reading the documentation
everybody that worked on https://github.com/scylladb/scylladb/issues/13276
didn't even suspect that the ALTER KEYSPACE below is a NOOP:
```
CREATE KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1 }
ALTER KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 2 }
```

**Comment by vladzcloudius** [comment]
> > Again, this is not confusing at all - this is a documented behavior.
> 
> This is not a legal argument. Let me rephrase that as counter-intuitive since without reading the documentation everybody that worked on #13276 didn't even suspect that the ALTER KEYSPACE below is a NOOP:
> 
> ```
> CREATE KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1 }
> ALTER KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 2 }
> ```

The above looks like a copy-paste bug in the test. Using 'replication_factor' while ALTERing NetworkStrategy is unorthodox to begin with. And whoever **intentionally** does that is expected to figure the semantics of this from the documentation and not rely on intuition.

**Comment by bhalevy** [comment]
> > > Again, this is not confusing at all - this is a documented behavior.
> > 
> > 
> > This is not a legal argument. Let me rephrase that as counter-intuitive since without reading the documentation everybody that worked on #13276 didn't even suspect that the ALTER KEYSPACE below is a NOOP:
> > ```
> > CREATE KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1 }
> > ALTER KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 2 }
> > ```
> 
> The above looks like a copy-paste bug in the test. Using 'replication_factor' while ALTERing NetworkStrategy is unorthodox to begin with. And whoever **intentionally** does that is expected to figure the semantics of this from the documentation and not rely on intuition.

Having a system where one needs to deep-dive into the documentation to figure out what exactly this seemingly straight-forward query does is really bad for adoption.  You want the users to be successful in developing applications with as short ramp-up as possible.  Simple stuff should stay simple, and complicated stuff should be made possible with coherent APIs and good documentation.

The incoherence of "ALTER KEYSPACE" vs. "CREATE KEYSPACE" is insane IMO.
CREATE KEYSPACE does the most expected thing - that's great.
but ALTER KEYSPACE has small print to it?

The alter statement says in plain English: "I want to change the replication factor for this keyspace".
It's a simple request.  If the system can do, great.  If not, it must return an error.
Applying this option implicitly only to data centers currently with rf=0 is really pulled out of thin air
and you probably need to the engineer that programmed that to understand why...

**Comment by nyh** [comment]
> @nyh @roydahan were you aware of that doc page? If you read it would it still be "confusing"?

As I explained in #8881 when I opened it already a couple of years ago, I first tried an ALTER TABLE and noticed it didn't do anything, so I went to the documentation, and indeed discovered that this "not doing anything" was an intentional and documented behavior - and even linked to the document in the issue. But I was lucky that I noticed the ALTER TABLE didn't do anything and started to read documentation - @bhalevy was less lucky and didn't notice, and the application simply did the wrong thing and nobody knew it.

> IMO there is nothing confusing about it. Nor I heard customers complaining about it either.

Don't you find it confusing that you do an ALTER TABLE hoping to change the replication factor, and it doesn't do anything at all, and doesn't tell you it didn't do anything?

I guess that changing the replication factor of an existing table is a rare operation, even more so with multi DC, and even more so combined with the (misnamed) "auto expand" feature - so we don't have many customers who actually ran into this. Both I and Benny ran into this issue because we were writing tests - not because we really wanted to use this feature for a real workload.

**Comment by nyh** [comment]
> > > Again, this is not confusing at all - this is a documented behavior.
> > 
> > 
> > This is not a legal argument. Let me rephrase that as counter-intuitive since without reading the documentation everybody that worked on #13276 didn't even suspect that the ALTER KEYSPACE below is a NOOP:
> > ```
> > CREATE KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1 }
> > ALTER KEYSPACE name WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 2 }
> > ```
> 
> The above looks like a copy-paste bug in the test. Using 'replication_factor' while ALTERing NetworkStrategy is unorthodox to begin with. And whoever **intentionally** does that is expected to figure the semantics of this from the documentation and not rely on intuition.

@vladzcloudius I think you're missing something... Traditionally, NetworkTopologyStrategy didn't support "replication_factor" at all - you needed to specify a replication factor separately per DC. Then, we added a feature that we (mis)named "auto expand" which allows you to specify replication_factor just once and it applies to all DCs. After you learn that it works for CREATE KEYSPACE, you can naturally assume that it also works for ALTER KEYSPACE - just like any other parameter you can set in CREATE KEYSPACE and later change with ALTER KEYSPACE. And it does seeminly "work" - it's not an error - it just does a very weird thing - it doesn't alter any existing DC's replication factor and only modifies replication factors on "new" DCs (one on which there was no replication yet).

**Comment by vladzcloudius** [comment]
> As I explained in https://github.com/scylladb/scylladb/issues/8881 when I opened it already a couple of years ago, I first tried an ALTER TABLE and noticed it didn't do anything, so I went to the documentation, and indeed discovered that this "not doing anything" was an intentional and documented behavior - and even linked to the document in the issue. 

Got it. So, at least it wasn't a surprise.

> But I was lucky that I noticed the ALTER TABLE didn't do anything and started to read documentation - @bhalevy was less lucky and didn't notice, and the application simply did the wrong thing and nobody knew it.

I think reading the documentation has nothing to do with luck. 

> @vladzcloudius I think you're missing something... Traditionally, NetworkTopologyStrategy didn't support "replication_factor" at all - you needed to specify a replication factor separately per DC. Then, we added a feature that we (mis)named "auto expand" which allows you to specify replication_factor just once and it applies to all DCs. After you learn that it works for CREATE KEYSPACE, you can naturally assume that it also works for ALTER KEYSPACE - just like any other parameter you can set in CREATE KEYSPACE and later change with ALTER KEYSPACE. And it does seeminly "work" - it's not an error - it just does a very weird thing - it doesn't alter any existing DC's replication factor and only modifies replication factors on "new" DCs (one on which there was no replication yet).

Thanks for clarifying. But allow me to add on top of that you wrote: I'm not sure it was "us" adding it and not "us following DataStax": https://cassandra.apache.org/doc/latest/cassandra/cql/ddl.html#networktopologystrategy

And the documentation text is surprisingly similar too (was it them copying it from us?).

Regardless who copied from who the current implementation is an ABI and unless there is a **really good** reason to change it (e.g. a security concern) we should not change it.

If not for these reasons and if I was the one who was defining this new semantics I'd keep it simple and would **always** change all DCs to a given RF - exactly as `CREATE KEYSPACE` does.

But, as I wrote, this train is left and we can't just change the existing user facing API simply because somebody finds it "confusing".

**Comment by denesb** [comment]
> > Don't we have a way to return warnings to the cql client? That way we could warn instead of silently ignoring the parameter, without breaking the "ABI".
> 
> IMO there is. But why would you want to give a warning when "2+0" comes out "2"? Some may call it "surprising" too but those who know semantics of a "+" operator - won't. And for them this warning would be an annoying noise.

The authors of compilers thought this is worthwhile, I saw the compiler issue warnings/errors like (not literal quote): 
* "statement X has no effect"
* "statement is always true"

The programmer should know, so why warn? The reason is that the programmer might have forgot. They might have read the docs at one point, but maybe not all of it, or maybe they didn't even bother. So it is better to be explicit and avoid surprises.

**Comment by bhalevy** [comment]
In addition to the above, the usage of the `replication_strategy` option in `ALTER KEYSPACE` depends on the replication strategy.  For the trivial case of a single DC, `SimpleStrategy` changes that DCs replication factor, while with `NetorkTopologyStrategy` the replication factor does not change.
A simpler and easier to internalize model would apply `replication_strategy` in this trivial case the same way.

Following the same principle, if all DCs in the cluster are set to the same RF, the replication_factor option can be corollary applied.  Since with the **current** semantics, applications have no reason to do that on purpose since it's a NOOP, I think that the above semantic change is safe as it augments the interface in a backward compatible way and can easily be explained in the documentation.

In the cases where some DCs have non-zero RF and some DCs have RF=0 (as in the scenario of adding a new DC to an existing cluster), I suggested here to return an error, which would break the interface as noted, but since those semantics are common to other vendors and we want to maintain compatibility, warning about it may be enough to encourage the user to set the RF on new DCs using an explicit option that is non-ambiguous.

**Comment by bhalevy** [comment]
Closing for noew as it was decided not to change the API
